third_party/ruby/penelope_procc.patch

```diff
@@ -0,0 +1,97 @@
+diff --git proc.c proc.c
+index 0c8889f04b..0f231efa8b 100644
+--- proc.c
++++ proc.c
+@@ -43,6 +43,9 @@ VALUE rb_cMethod;
+ VALUE rb_cBinding;
+ VALUE rb_cProc;
+ 
++static int ME_CLONE_DO_NOT_CLONE = false;
++static int ME_CLONE_CLONE = true;
++
+ static rb_block_call_func bmcall;
+ static int method_arity(VALUE);
+ static int method_min_max_arity(VALUE, int *max);
+@@ -2367,7 +2370,7 @@ rb_method_call_with_block(int argc, const VALUE *argv, VALUE method, VALUE passe
+  */
+ 
+ static void
+-convert_umethod_to_method_components(VALUE method, VALUE recv, VALUE *methclass_out, VALUE *klass_out, VALUE *iclass_out, const rb_method_entry_t **me_out)
++convert_umethod_to_method_components(VALUE method, VALUE recv, VALUE *methclass_out, VALUE *klass_out, VALUE *iclass_out, const rb_method_entry_t **me_out, bool clone)
+ {
+     struct METHOD *data;
+ 
+@@ -2389,9 +2392,19 @@ convert_umethod_to_method_components(VALUE method, VALUE recv, VALUE *methclass_
+ 	}
+     }
+ 
+-    const rb_method_entry_t *me = rb_method_entry_clone(data->me);
++    const rb_method_entry_t *me;
++
++    if (clone == ME_CLONE_CLONE) {
++	me = rb_method_entry_clone(data->me);
++    } else {
++	me = data->me;
++    }
++   
+ 
+     if (RB_TYPE_P(me->owner, T_MODULE)) {
++	if (clone == ME_CLONE_DO_NOT_CLONE) {
++	    me = rb_method_entry_clone(data->me);
++	}
```

Comment:
Why does this check that `clone` is "do not clone" but then immediately call `*_clone`?

Reply:
in this branch, we want to always, unconditionally clone the method entry, because things lower down manipulate me

Reply:
this maps to this stanza in my ruby pr:https://github.com/ruby/ruby/pull/6244/files#diff-b2553d23e6b1fe76e20608d06c25f6acca06279100f1a9c24febcd79a82fac3cR2576-R2580